### PR TITLE
Input vectors

### PR DESCRIPTION
**Pull request by jsenellart-systran**
Added 2 new features for adding support of input vectors
- option `-idx_files` indicate that source/target files are indexed meaning that their format is `key value` not necessarily in the same order
- new data type `feattext` which is using Kaldi text ark dump format 

```
KEY [
VAL1 VAL2 VAL3
VAL4 VAL5 VAL6 ]
KEY2 [
...
]
```

Typical preprocessing command:
```
th preprocess.lua -data_type 'feattext' -train_src TIMIT/raw_mfcc_train.ark.txt \
                -train_tgt TIMIT/data/train/text -valid_src TIMIT/raw_mfcc_dev.ark.txt \
                -valid_tgt TIMIT/data/dev/text -save_data TIMIT/baseline -idx_files \
                -src_seq_length 1000 -tgt_seq_length 100
```

For training:
```
th train.lua -data TIMIT/baseline-train.t7 -save_model TIMIT -pdbrnn -report_every 1 \
           -rnn_size 800 -word_vec_size 20  -layers 4 -max_batch_size 16 -learning_rate 0.7 \
           -learning_rate_decay 0.8 -end_epoch 20
```

For decoding:
```
th translate.lua -model data/asr1_epoch15_8.17.t7 -src TIMIT/raw_mfcc_test.ark.txt  -batch_size 1
```

### CODE REVIEW

File: onmt/modules/PDBiEncoder.lua [hunk 1]
```diff
@@ -110,6 +110,23 @@ function PDBiEncoder:maskPadding()
   self.layers[1]:maskPadding()
 end
 
+-- size of context vector
+function PDBiEncoder:contextSize(sourceSize, sourceLength)
```
Comment (jsenellart-systran) [inline onmt/modules/PDBiEncoder.lua]: contextSize is used in decoder to have size of the context size outside of the encoder which varies for pdbiencoder

File: preprocess.lua [hunk 2]
```diff
@@ -57,28 +60,48 @@ local function main()
...
-                                   opt.features_vocabs_prefix,
-                                   function(s) return isValid(s, opt.src_seq_length or opt.seq_length) end)
+
+  _G.logger:info('Preparing vocabulary...')
+  if dataType ~= 'feattext' then
+    local src_file = opt.train_src
+    if dataType == 'monotext' then
+      src_file = opt.train
+    end
+    data.dicts.src = Vocabulary.init('train',
```
Comment (guillaumekln) [inline preprocess.lua]: Can we keep `'source'` instead? See https://github.com/OpenNMT/OpenNMT/pull/162.

File: onmt/data/Batch.lua [hunk 3]
```diff
@@ -74,14 +74,24 @@ function Batch:__init(src, srcFeatures, tgt, tgtFeatures)
 
   self.sourceLength, self.sourceSize, self.uneven = getLength(src)
 
+  -- if input vectors (speech for instance)
+  self.inputVectors = src[1]:dim() > 1
```
Comment (geffy) [inline onmt/data/Batch.lua]: This fails with default constructor like `local batch = Batch()`

Comment (guillaumekln) [inline onmt/data/Batch.lua]: Good catch! It has been fixed.

**Pull request merged**